src/actions/tamanu/prepare_upgrade.rs

```diff
@@ -0,0 +1,160 @@
+
+	/// Force installing older Tamanu
+	#[arg(long)]
+	pub force: bool,
+}
+
+pub async fn run(ctx: Context<TamanuArgs, PrepareUpgrade>) -> Result<()> {
+	let PrepareUpgrade {
+		version: new_version,
+		package: _,
+		force,
+	} = ctx.args_sub;
+
+	let (_, root) = find_tamanu(&ctx.args_top)?;
+	let existing_version = find_existing_version()?;
+
+	let package = match ctx.args_sub.package {
+		Some(package) => package,
+		None => find_package(&root)?,
+	};
+	info!(?package, "using");
+	let kind = match package.as_str() {
+		"central-server" => ServerKind::Central,
+		"facility-server" => ServerKind::Facility,
+		_ => bail!("package {package} not recognised"),
+	};
+
+	// FIXME: this may not support for Linux as directory structures are different
+	// `root` should have a parent.
+	let existing_root = root
+		.parent()
+		.unwrap()
+		.join(format!("release-v{existing_version}"));
+	let new_root = root
+		.parent()
+		.unwrap()
+		.join(format!("release-v{new_version}"));
+	let new_web_root = root
+		.parent()
+		.unwrap()
+		.join(format!("tamanu-web-{new_version}"));
+
+	let minimum_version = Version::parse("2.0.0").unwrap();
+	if existing_version < minimum_version || new_version < minimum_version {
+		bail!("`PreUpgrade-Tamanu` only support upgrading from/to versions from 2.0.0 onwards");
+	}
+
+	if !force && (new_version <= existing_version) {
+		bail!("the specified version is older than or equal to the installed version");
+	}
+
+	if !new_root.exists() {
+		let url = make_url(kind, new_version.to_string())?;
+		// FIXME: this may not support for Linux as directory structures are different
+		download(url, root.parent().unwrap()).await?;
+	}
+
+	if !new_web_root.exists() {
+		let url = make_url(ServerKind::Web, new_version.to_string())?;
+		// FIXME: this may not support for Linux as directory structures are different
+		download(url, root.parent().unwrap()).await?;
+	}
+
+	duct::cmd!("pwsh", "-Command", "yarn")
+		.dir(&new_root)
+		.run()
+		.into_diagnostic()
+		.wrap_err("failed to run yarn")?;
+
+	let config_path = ["packages", &package, "config", "local.json5"]
+		.iter()
+		.collect::<PathBuf>();
+	let existing_config = existing_root.join(&config_path);
+	let new_config = new_root.join(&config_path);
+	info!(config = ?existing_config.display(), "copying configs to the new Tamanu installation");
+	fs::copy(existing_config, new_config).into_diagnostic()?;
+
+	info!("checking the new version is runnable");
+	duct::cmd!("node", "dist", "help")
+		.dir(&new_root.join("packages").join(&package))
+		.run()
+		.into_diagnostic()?;
+
+	if has_non_deterministic_migrations(&existing_root, &new_root)? {
+		println!("Warning: the upgrade may contain (a) non-deterministic migration(s).");
+	}
+
+	Ok(())
+}
+
+/// Checks if there's new migration since the last update.
+fn has_non_deterministic_migrations(
+	existing_root: impl AsRef<Path>,
+	new_root: impl AsRef<Path>,
+) -> Result<bool> {
+	let re = Regex::new(r"NON_DETERMINISTIC += +true").unwrap();
```

Comment:
```suggestion
	let re = Regex::new(r"NON_DETERMINISTIC\s*=\s*true").unwrap();
```

Just to cover different kinds of whitespace.